transform_aes.c

```diff
@@ -259,37 +245,80 @@ static int transop_decode_aes( n2n_trans_op_t * arg,
 }
 
 static int setup_aes_key(transop_aes_t *priv, const uint8_t *key, ssize_t key_size) {
-    size_t aes_keysize_bytes;
-    size_t aes_keysize_bits;
-    struct sha512_keybuf keybuf;
+    size_t aes_key_size_bytes;
+    size_t aes_key_size_bits;
+    uint8_t * key_mat_buf;
+    size_t key_mat_buf_length;
 
     /* Clear out any old possibly longer key matter. */
     memset( &(priv->enc_key), 0, sizeof(priv->enc_key) );
     memset( &(priv->dec_key), 0, sizeof(priv->dec_key) );
     memset( &(priv->iv_enc_key), 0, sizeof(priv->iv_enc_key) );
-    memset( &(priv->iv_ext_val), 0, sizeof(priv->iv_ext_val) );
+    memset( &(priv->iv_pad_val), 0, sizeof(priv->iv_pad_val) );
+
+    /* Let the user choose the degree of encryption:
+     * Long input keys will pick AES192 or AES256 with more robust but expensive encryption.
+     *
+     * The input key always gets hashed to make a more unpredictable use of the key space and
+     * also to derive some additional material (key for IV encrpytion, IV padding).
+     *
+     * The following scheme for key setup was discussed on github: 
+     * https://github.com/ntop/n2n/issues/101
+     */
 
-    /* We still use aes_best_keysize (even not necessary since we hash the key
-     * into the 256bits enc_dec_key) to let the users choose the degree of encryption.
-     * Long keys will pick AES192 or AES256 with more robust but expensive encryption.
+    /* create a working buffer of maximal occuring hashes size and generate
+     * the hashes for the aes key material, key_mat_buf_lengh indicates the
+     * actual "filling level" of the buffer
      */
-    aes_keysize_bytes = aes_best_keysize(key_size);
-    aes_keysize_bits = 8 * aes_keysize_bytes;
+    key_mat_buf = calloc(1, SHA512_DIGEST_LENGTH + SHA256_DIGEST_LENGTH);
```

Comment:
A normal stack allocation with `uint8_t key_mat_buf[SHA512_DIGEST_LENGTH + SHA256_DIGEST_LENGTH]` should be enough

Reply:
You are right, that would be much easier... and doesn't create the need to take care of de-allocation.